### PR TITLE
GEODE-6730: optimize check for sender id differences

### PR DESCRIPTION
**Pull request by dschneider-pivotal**
Instead of every region write operation scanning the profiles and doing set comparisions to detect if any of the gateway or async event queue ids on a region are configured differently, now this check is done whenever the profiles or our local sender ids change. The region write operations now just check two volatiles to see if they are non-null.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java [hunk 1]
```diff
@@ -453,10 +466,21 @@ public boolean initializationGate() {
...
       return false;
     }
-    synchronized (initializeLock) {
-      if (!initialized) {
-        exchangeProfiles();
-        return true;
+    boolean exchangedProfiles = false;
+    try {
+      synchronized (initializeLock) {
+        if (!initialized) {
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java]: Could this sync be replaced with a compareAndSet on initialized?

Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java]: I don't think that would be safe. If two thread come into this at the same time, then we want the second thread to wait for the first thread to finish the initialization. And we only want one thread to do the profile exchange. I don't see a performance problem with this existing code. Initialization happens once early and after that this is just a volatile read of a boolean which I think would perform the same as a compareAndSet call. Let me know if I'm missing something.

**Comment by dschneider-pivotal** [comment]
> I like the performance improvement, and it looks like you are trying to get this check working without changing behavior, which is a good start.
> 
> I do wonder if we need to rethink how (or if) we should be validating this during a write operation. It seems like throwing this exception in the middle of the write operation (after cache update but before notifying any gateways) will result in data inconsistencies or at least inconsistent deliveries of events to gateways during the time window that someone is adding a new gateway. Is that even worse than not checking at all?

My first thought was the validation should be done at the time a new member joins or when a gateway or async event queue is added or removed. But the problem with the latter is that it is not a distributed operation so the user must add or remove it to multiple members manually. We could have the current check done by the write ops wait for a while when a difference is detected in hopes that the user is working on making it consistent. I asked @boglesby to take a look at this. Perhaps we need another ticket that does focus on performance but on how and when sender id consistency should be checked.

**Comment by boglesby** [comment]
I think we talked about removing that check back when we were doing the lucene work. Destroying a lucene index removes an AEQ id from the region, so a put was failing in the checkSameSenderIdsAvailableOnAllNodes call in notifyGatewaySender. We ended up adding the visibleAsyncEventQueueIds set and only checking that. Lucene AEQ ids don't get added to that set.

This issue is definitely in there for normal AEQ ids, though. One saving grace is that generally customers don't add AEQs to a region on the fly. They are there from the beginning.

Having that check on every put operation seems like overkill. I think it only has to be checked when a new member joins and when a new GatewaySender or AEQ is added or removed not using gfsh.

When a new member joins with different AEQ ids, this exception is thrown:

[error 2019/05/03 16:30:47.673 PDT <main> tid=0x1] java.lang.IllegalStateException: Cannot create Region /data with [ db2, db] async event ids because another cache has the same region defined with [db] async event ids
```
Exception in thread "main" java.lang.IllegalStateException: Cannot create Region /data with [ db2, db] async event ids because another cache has the same region defined with [db] async event ids
	at org.apache.geode.internal.cache.CreateRegionProcessor$CreateRegionMessage.handleCacheDistributionAdvisee(CreateRegionProcessor.java:466)
	at org.apache.geode.internal.cache.CreateRegionProcessor$CreateRegionMessage.process(CreateRegionProcessor.java:369)
```
So, that case is already covered.

Currently when an existing member adds an AEQ id that isn't defined in another member, no exception is thrown. The exception happens when a put occurs like:
```
Caused by: org.apache.geode.internal.cache.wan.GatewaySenderConfigurationException: Region data has [db] AsyncEvent queue IDs. Another cache has same region with [db2, db] AsyncEvent queue IDs. For region across all members, AsyncEvent queue IDs should be same.
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.PartitionedRegion.checkSameSenderIdsAvailableOnAllNodes(PartitionedRegion.java:1295)
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.LocalRegion.notifyGatewaySender(LocalRegion.java:6217)
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.BucketRegion.notifyGatewaySender(BucketRegion.java:716)
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.LocalRegion.basicPutPart2(LocalRegion.java:5863)
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.BucketRegion.basicPutPart2(BucketRegion.java:701)
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.map.RegionMapPut.doBeforeCompletionActions(RegionMapPut.java:282)
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.AbstractRegionMap.basicPut(AbstractRegionMap.java:2044)
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.BucketRegion.virtualPut(BucketRegion.java:527)
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.LocalRegionDataView.putEntry(LocalRegionDataView.java:162)
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.LocalRegion.basicUpdate(LocalRegion.java:5672)
	at Remote Member '192.168.1.2(25386)<v1>:41001' in org.apache.geode.internal.cache.AbstractUpdateOperation.doPutOrCreate(AbstractUpdateOperation.java:150)
	at org.apache.geode.internal.cache.BucketRegion.basicPutPart2(BucketRegion.java:691)
	at org.apache.geode.internal.cache.map.RegionMapPut.doBeforeCompletionActions(RegionMapPut.java:282)
	at org.apache.geode.internal.cache.tier.sockets.command.Put65.cmdExecute(Put65.java:392)
```
In that case, the region has the entry, but it wasn't sent to the AEQ.

**Comment by dschneider-pivotal** [comment]
After discussions with Dan and Barry, we decided that they best way to optimize this check is to not do the check. Currently the action taken by the check causes data inconsistency. So I'm going to close this pull request and submit another one that removes the check.

**Pull request closed without merging**